Create script for automated regression testing

ilectra opened the pull request:
Fixes #8 .
~This is stil WIP, I need to~
- [x] check that all files are copied properly to the `build` directory and paths are correct
- [x] possibly convert to `pytest`

But you can already use it from inside the directory that the test sits in `build`, ~after copying the file with the reference values over.~
@edbennett are the values we're checking the right ones? Or do we need the palette value to a gazillion decimals?

ilectra committed:
First attempt at regression test script.

qiUip commented on tests/run_regression_test.py:
```diff
@@ -0,0 +1,134 @@
+    result = compare(plaquette, expected_plaquette, "plaquette", args.stop)
+    result = result and compare(checksum_rng, expected_checksum_rng, "Checksum RNG file ", args.stop)
+    result = result and compare(checksum_lat, expected_checksum_lat, "Checksum LAT file ", args.stop)
+    if result:
+        print("All tests passed!")
+    else:
+        print("Some tests failed...")
+
+    os.remove("output.txt")
+    os.remove("ckpoint_rng.1")
+    os.remove("ckpoint_lat.1")
```
```suggestion
    if result:
        print("All tests passed!")
        os.remove("output.txt")
    else:
        print("Some tests failed...")

    os.remove("ckpoint_rng.1")
    os.remove("ckpoint_lat.1")
```

qiUip replied:
It might be useful not to remove the output if the test fails?

edbennett replied:
I'd also anticipate a non-zero exit code for a failed test? (Obviously converting to pytest would give you that for free)

ilectra replied:
Good call. I committed this for now, and will see how to best convert to pytest.

qiUip commented on tests/run_regression_test.py:
```diff
@@ -0,0 +1,134 @@
+
+    expected_plaquette, expected_checksum_rng, expected_checksum_lat = read_expected(args.test_name, args.grid, args.mpi)
+
+    result = subprocess.run([f"./{args.test_name} --grid {args.grid} --mpi {args.mpi} --Thermalizations 0 --Trajectories 1 > output.txt"], shell=True, encoding="text")
```
```suggestion
    result = subprocess.run([f"./{args.test_name} --grid {args.grid} --mpi {args.mpi} --Thermalizations 0 --Trajectories 1 --threads 1 > output.txt"], shell=True, encoding="text")
```

qiUip replied:
Without enforcing the threads, I got different results (rounded up) so probably best to enforce it as default and maybe parameterise threads as a potential option.

edbennett replied:
To confirm, seeing different results for the HMC tests based on gauge group is expected, as you're looking at different physics.

qiUip replied:
Good to have confirmation - this just means that the expected results are not only "for the test" but for a specific gauge group.  Just something that needs to be clear / explained somewhere so that we can track which expected results matches which setup.

qiUip commented on tests/sp2n/Test_hmc_Sp_WilsonFundFermionGauge_expected.txt:
```diff
@@ -0,0 +1 @@
+8.8.8.8 1.1.1.1 0.0256253844 922c392f d1e4cc1c
```
```suggestion
8.8.8.8 1.1.1.1 0.281507042 922c392f 716ec776
```

qiUip replied:
Updated expected output to align with original test (MDsteps = 36 and trajL = 1.0).

ilectra replied:
Would we want to keep both test cases? ((36,1.0) and (2,0.1))

ilectra replied:
Yes, we want to keep both `MDsteps` and `trajL` as test parameters.

qiUip commented on tests/sp2n/Test_hmc_Sp_WilsonFundFermionGauge.cc:
```diff
@@ -64,8 +64,8 @@ int main(int argc, char **argv) {
     
-  TheHMC.Parameters.MD.MDsteps = 36;
-  TheHMC.Parameters.MD.trajL   = 1.0;
+  TheHMC.Parameters.MD.MDsteps = 2;
+  TheHMC.Parameters.MD.trajL   = 0.1;
```
```suggestion
  TheHMC.Parameters.MD.MDsteps = 36;
  TheHMC.Parameters.MD.trajL   = 1.0;
```

qiUip replied:
I think it would be best to to change the test case in the develop branch as it should stay consistent with the upstream.

ilectra replied:
I hope you mean "not to change the test case...." :) . The problem is that the expected value is for the modified `MDsteps` and `trajL`. Do we want to ditch them completely? (I saw you fixed those in the expected values file below)

Commit:
Adding expected values for 2 additional SU3 tests

qiUip reviewed (changes requested):
In terms of functionality, this works great! A few suggestions in the review.

I also noticed that the results change based on compilation options, e.g. when compiling for sp2n (`--enable-Sp --enable-Nc=4`) and running Test_hmc_IwasakiGauge I got different results than the same test when compiled for su3 (`--enable-Nc=3`).  I'm not sure these values will make a lot of sense without a very strict compile and run parameters.

edbennett commented:
Something I should have considered earlier: since this uses a Metropolis algorithm, checking the final value of the plaquette is only valid if the version being compared against accepts the change. (In the case of a rejected change, then the value of the plaquette output is the input value, and only depends on the code we are touching if it changes in such a way that the change is accepted instead, which is unlikely.)

Provided we consistently stick with the same test case (and have verified that it accepts), then I think this is fine, but we should be careful about generalising this.

ilectra committed:
Apply suggestions from code review

Co-authored-by: Mashy Green <mashy.green@ucl.ac.uk>

ilectra committed:
Initial port to pytest - messy!

ilectra committed:
Clean pytest implementation.

ilectra commented:
- [x] To add more parameters/references to check for: nthreads, CPUvsGPU
- [ ] ~read references from command line (if possible)~ I don't think this is a good idea. It should be easy to add a line in the `...._expected.txt` file.

ilectra committed:
Add function to read expected values by line.
SOme improvements in error checking and reporting.

ilectra committed:
Read expected values from file by line.

ilectra committed:
Read test parameters as well from the file, remove from command line.

ilectra committed:
Link over needed files for tests with autoconf.
Future test reference value files added, will have to be linked over explicitly here as well. autotools does not do wildcards.

ilectra commented:
To be able to tell if the run was with GPU or CPU, look for `nvidia::memalloc` or `unified memory` in the output.

ilectra committed:
Add nthreads as test parameter

ilectra committed:
Add the rest of the requested test parameters.

ilectra committed:
Keep MDsteps and trajL in the test as they were

Co-authored-by: Mashy Green <mashy.green@ucl.ac.uk>

ilectra commented:
I think this is ready for a final review. Please have a look and add test cases values that you find useful to the `Test_hmc_Sp_WilsonFundFermionGauge_expected.txt` file. @qiUip @edbennett @asifsamiarain

ilectra commented on tests/hmc/Test_hmc_IwasakiGauge_expected.txt:
```diff
@@ -0,0 +1 @@
+8.8.8.8 1.1.1.1 0.269298793 633bf471 3a22ad20
```
Those two expected values files have to be adjusted for the extra test parameters and line numbers. @qiUip

ilectra commented:
Add usage note to script

edbennett commented:
As you noted, a note on usage would be useful. You might want to run `ruff format` or similar to tidy up some very minor things. Other than that, looks good to me!

ilectra committed:
Add documentation to run_regression_tests script. Various debugging.

ilectra committed:
Add another test case to Test_hmc_Sp_WilsonFundFermionGauge

ilectra committed:
Add the hmc tests to configure.ac.
Add some more test parameters and expected values - in progress...
Modify saveInterval in hmc tests to produce outputs.
Fix number of threads to 1 if not specified.

ilectra committed:
Merge upstream and local develop

ilectra commented:
The `lat.checkoint` file checksum is machine dependent, therefore not appropriate for regression tests (unless we keep note of the machine where the reference values were generated as well). A better value to use for reference is `LINK_TRACE`, found in line 9 of the `lat.checkpoint` header.
Do we still want to keep the checksum for more detailed comparisons? In this case, the user would have to generate the checksum before the changes in the code they want to check for exact bit-wise identity, and then compare it with the one after. A command line option for the pytest, eg. `--checksum`, could take no argument to just report the checksum after the run, or take the value to compare with as an argument. Either way, it'll be a 2-step process, where the user will have to run pytest with the old version of the code to generate the checksum value, then re-compile with the new version of the code and run pytest again to do the comparison. Does this seem worth the effort @edbennett ?

edbennett commented:
Having a tool that makes it easier to do a checksum comparison before and after a change, so you can verify that a trajectory is bitwise compatible when run on the same machine, would be useful. Whether it's worth the effort of course depends on how much effort it would be. I'd say it's worth a single-digit number of hours of effort, but not that much more than that.
